composer/trainer/trainer.py

```diff
@@ -1325,6 +1353,23 @@ def __init__(
             # Only wrap the module if required
             self.state.model = prepare_ddp_module(self.state.model, self._find_unused_parameters)
 
+        # The model would need to be torch.compile()'d after being wrapped in a distributed strategy
+        # to take advantage of any graph breaks.
+        if is_torch_2_0 and not is_model_compiled and compile_config is not None:
+            compiled_model = torch.compile(  # pyright: ignore [reportGeneralTypeIssues]
+                self.state.model, **compile_config)
+            self.state.model = compiled_model._orig_mod
+            self.state.model.forward = compiled_model.dynamo_ctx(self.state.model.forward)
+            is_model_compiled = True
+            # update local_hparams to ensure the `is_model_compiled` is set correctly for
+            # debugging purpose and for unit test.
+            if self.auto_log_hparams:
+                self.local_hparams['is_model_compiled'] = is_model_compiled
+        elif not is_torch_2_0 and compile_config is not None:
+            warnings.warn(f'`torch.compile` is supported for PyTorch 2.0 or higher.' +
+                          f'Either update your PyTorch version or disable parameter by providing ' +
+                          f'`compile_config` to `None`.')
```

Comment:
Should this be something like a ValueError and not just a warning?

Reply:
I have kept warning message since we do not want to break the flow if user by mistakenly provided compile_config for torch < 2.0 and the training can still proceed even if we don't compile the model.